compiler/semantic/sql.go

```diff
@@ -16,61 +16,292 @@ import (
+// and also sort by the out elements.  It's up to the caller to unwrap the in/out
+// record when returning to pipeline context.
+func (a *analyzer) semSelect(sel *ast.Select, seq dag.Seq) (dag.Seq, schema) {
+	if len(sel.Selection.Args) == 0 {
+		a.error(sel, errors.New("SELECT clause has no selection"))
+		return seq, badSchema()
+	}
+	var fromSchema schema
+	seq, fromSchema = a.semSelectFrom(sel.Loc, sel.From, seq)
+	if fromSchema == nil {
+		return seq, badSchema()
+	}
+	if sel.Value {
+		return a.semSelectValue(sel, fromSchema, seq)
+	}
+	sch := &schemaSelect{in: fromSchema}
+	var funcs aggfuncs
+	proj := a.semProjection(sch, sel.Selection.Args, &funcs)
+	var where dag.Expr
+	if sel.Where != nil {
+		where = a.semExprSchema(sch, sel.Where)
+	}
+	keyExprs := a.semGroupBy(sch, sel.GroupBy)
+	having, err := a.semHaving(sch, sel.Having, &funcs)
+	if err != nil {
+		a.error(sel.Having, err)
+		return seq, badSchema()
+	}
+	// Now that all the pieces have been converted to DAG fragments,
+	// we stitch together the fragments into pipeline operators depending
+	// on whether its an aggregation or a selection of scalar expressions.
+	seq = yieldExpr(wrapThis("in"), seq)
+	var schOut schema
+	if len(funcs) != 0 || len(keyExprs) != 0 {
+		seq = a.genSummarize(sel.Loc, proj, where, keyExprs, funcs, having, seq)
+		schOut = sch.out
+	} else {
+		if having != nil {
+			a.error(sel.Having, errors.New("HAVING clause requires aggregations and/or a GROUP BY clause"))
 		}
-		// If we have parents with both a from and select, report an error but
-		// only if it's not a RobotScan where the parent feeds the from operateor.
-		if _, ok := seq[off].(*dag.RobotScan); !ok {
-			if hasParent {
-				a.error(sel, errors.New("SELECT cannot have both an embedded FROM claue and input from parents"))
-				return append(seq, badOp())
-			}
+		seq = a.genYield(proj, where, sch, seq)
+		schOut = sch
+		if sel.Distinct {
+			seq = a.genDistinct(pathOf("out"), seq)
 		}
 	}
-	if sel.Value {
-		return a.semSelectValue(sel, seq)
+	return seq, schOut
+}
+
+func (a *analyzer) semHaving(sch schema, e ast.Expr, funcs *aggfuncs) (dag.Expr, error) {
+	if e == nil {
+		return nil, nil
 	}
-	proj, ok := a.semProjection(sel.Selection.Args)
-	if !ok {
-		return dag.Seq{badOp()}
+	return funcs.subst(a.semExprSchema(sch, e))
+}
+
+func (a *analyzer) genYield(proj projection, where dag.Expr, sch *schemaSelect, seq dag.Seq) dag.Seq {
+	if len(proj) == 0 {
+		return nil
 	}
-	if sel.Where != nil {
-		seq = append(seq, dag.NewFilter(a.semExpr(sel.Where)))
+	seq = a.genColumns(proj, sch, seq)
+	if where != nil {
+		seq = append(seq, dag.NewFilter(where))
 	}
-	if sel.GroupBy != nil {
-		if proj.hasStar() {
-			a.error(sel, errors.New("aggregate mixed with *-selector not yet supported"))
-			return append(seq, badOp())
+	return seq
+}
+
+func (a *analyzer) genColumns(proj projection, sch *schemaSelect, seq dag.Seq) dag.Seq {
+	notFirst := false
```

Comment:
Nit: Use `var` for the zero value.
```suggestion
	var notFirst bool
```